comms/src/message/envelope.rs

```diff
@@ -81,15 +139,16 @@ impl MessageEnvelope {
             body = encrypt_envelope_body(&node_identity.secret_key, &dest_public_key, &body)?
         }
 
-        let signature = crypto::sign(&mut OsRng::new().unwrap(), node_identity.secret_key.clone(), &body)
-            .map_err(MessageError::SchnorrSignatureError)?;
-        let signature = signature.to_binary().map_err(MessageError::MessageFormatError)?;
+        let origin_signature = origin_signature(node_identity, dest.clone(), body.clone())?;
+        let peer_signature = peer_signature(node_identity, origin_signature.clone(), body.clone())?;
```

Comment:
Do you have to clone the body here as well?
Cloning a body message seems expensive?

Reply:
Removing the clones is difficult because of the way the challenges need to be generated, two or more bytes streams need to be combined to generate the signatures. I'm sure there might be better ways to do this, but I think it might have to be addressed in a separate issue.